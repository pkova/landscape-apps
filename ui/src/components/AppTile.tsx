import { Charge } from '@urbit/api'
import React, { useState } from 'react'

function normalizeUrbitColor(color: string): string {
    if (color.startsWith('#')) {
        return color
    }
    return `#${color.slice(2).replace('.', '').toUpperCase()}`
}

export default function AppTile({
    title,
    image,
    color,
}: Pick<Charge, 'title' | 'image' | 'color'>) {
    const [imageError, setImageError] = useState(false)

    return (
        <div
            className="overflow-hidden relative flex-none mr-3 w-12 h-12 bg-gray-200 rounded-lg"
            style={{ backgroundColor: normalizeUrbitColor(color) }}
        >
            {image && !imageError && (
                <img
                    className="object-cover absolute top-0 left-0 w-full h-full"
                    src={image}
                    alt={`app tile for ${title}`}
                    onError={() => setImageError(true)}
                />
            )}
        </div>
    )
}
