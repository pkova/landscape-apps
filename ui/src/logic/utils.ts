import ob from 'urbit-ob';
import { unixToDa } from '@urbit/api';
import { formatUv } from '@urbit/aura';
import anyAscii from 'any-ascii';
import { format, differenceInDays, endOfToday } from 'date-fns';
import _ from 'lodash';
import { Chat, ChatWhom } from '@/types/chat';
import {
  Cabals,
  GroupChannel,
  ChannelPrivacyType,
  Cordon,
  PrivacyType,
  Rank,
} from '@/types/groups';
import { CurioContent, Heap } from '@/types/heap';

export function nestToFlag(nest: string): [string, string] {
  const [app, ...rest] = nest.split('/');

  return [app, rest.join('/')];
}

export function renderRank(rank: Rank, plural = false) {
  if (rank === 'czar') {
    return plural ? 'Galaxies' : 'Galaxy';
  }
  if (rank === 'king') {
    return plural ? 'Stars' : 'Star';
  }
  if (rank === 'duke') {
    return plural ? 'Planets' : 'Planet';
  }
  if (rank === 'earl') {
    return plural ? 'Moons' : 'Moon';
  }
  return plural ? 'Comets' : 'Comet';
}

/**
 * Processes a string to make it `@tas` compatible
 */
export function strToSym(str: string): string {
  const ascii = anyAscii(str);
  return ascii.toLowerCase().replaceAll(/[^a-zA-Z0-9-]/g, '-');
}

export function channelHref(flag: string, ch: string) {
  return `/groups/${flag}/channels/${ch}`;
}

export function makePrettyTime(date: Date) {
  return format(date, 'HH:mm');
}

export function makePrettyDay(date: Date) {
  const diff = differenceInDays(endOfToday(), date);
  switch (diff) {
    case 0:
      return 'Today';
    case 1:
      return 'Yesterday';
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')}`;
  }
}

export function makePrettyDayAndTime(date: Date) {
  const diff = differenceInDays(endOfToday(), date);
  const time = makePrettyTime(date);
  switch (true) {
    case diff === 0:
      return `Today • ${time}`;
    case diff === 1:
      return `Yesterday • ${time}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time}`;
    default:
      return `${format(date, 'LLLL')} ${format(date, 'do')} • ${time}`;
  }
}

export function makePrettyDayAndDateAndTime(date: Date) {
  const diff = differenceInDays(endOfToday(), date);
  const time = makePrettyTime(date);
  const fullDate = `${format(date, 'LLLL')} ${format(date, 'do')}, ${format(
    date,
    'yyyy'
  )}`;
  switch (true) {
    case diff === 0:
      return `Today • ${time} • ${fullDate}`;
    case diff === 1:
      return `Yesterday • ${time} • ${fullDate}`;
    case diff > 1 && diff < 8:
      return `${format(date, 'cccc')} • ${time} • ${fullDate}`;
    default:
      return `${fullDate} • ${time}`;
  }
}

export function whomIsDm(whom: ChatWhom): boolean {
  return whom.startsWith('~') && !whom.match('/');
}

// ship + term, term being a @tas: lower-case letters, numbers, and hyphens
export function whomIsFlag(whom: ChatWhom): boolean {
  return (
    /^~[a-z-]+\/[a-z]+[a-z0-9-]*$/.test(whom) &&
    ob.isValidPatp(whom.split('/')[0])
  );
}

export function whomIsMultiDm(whom: ChatWhom): boolean {
  return whom.startsWith(`0v`);
}

export function normalizeUrbitColor(color: string): string {
  if (color.startsWith('#')) {
    return color;
  }

  const colorString = color.slice(2).replace('.', '').toUpperCase();
  const lengthAdjustedColor = _.padEnd(colorString, 6, _.last(colorString));
  return `#${lengthAdjustedColor}`;
}

export function pluralize(word: string, count: number): string {
  if (count === 1) {
    return word;
  }

  return `${word}s`;
}

export function createStorageKey(name: string): string {
  return `~${window.ship}/${window.desk}/${name}`;
}

// for purging storage with version updates
export function clearStorageMigration<T>() {
  return {} as T;
}

export const storageVersion = parseInt(
  import.meta.env.VITE_STORAGE_VERSION,
  10
);

export function preSig(ship: string): string {
  if (!ship) {
    return '';
  }

  if (ship.trim().startsWith('~')) {
    return ship.trim();
  }

  return '~'.concat(ship.trim());
}

export function newUv(seed = Date.now()) {
  return formatUv(unixToDa(seed));
}

export function getSectTitle(cabals: Cabals, sect: string) {
  return cabals[sect]?.meta.title || sect;
}

export function getFlagParts(flag: string) {
  const parts = flag.split('/');

  return {
    ship: parts[0],
    name: parts[1],
  };
}

export function getGroupPrivacy(cordon: Cordon): PrivacyType {
  if ('open' in cordon) {
    return 'public';
  }

  if ('shut' in cordon) {
    return 'private';
  }

  return 'secret';
}

export function getPrivacyFromChannel(
  groupChannel?: GroupChannel,
  channel?: Chat | Heap
): ChannelPrivacyType {
  if (!groupChannel || !channel) {
    return 'public';
  }

  if (groupChannel.readers.includes('admin')) {
    return 'secret';
  }

  if (channel.perms.writers.includes('admin')) {
    return 'read-only';
  }

  return 'public';
}

export function toTitleCase(s: string): string {
  if (!s) {
    return '';
  }
  return s
    .split(' ')
    .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
    .join(' ');
}

export function randomElement<T>(a: T[]) {
  return a[Math.floor(Math.random() * a.length)];
}

export function hasKeys(obj: Record<string, unknown>) {
  return Object.keys(obj).length > 0;
}

export const IMAGE_REGEX =
  /(\.jpg|\.img|\.png|\.gif|\.tiff|\.jpeg|\.webp|\.webm|\.svg)$/i;
export const AUDIO_REGEX = /(\.mp3|\.wav|\.ogg|\.m4a)$/i;
export const VIDEO_REGEX = /(\.mov|\.mp4|\.ogv)$/i;
export const URL_REGEX = /(https?:\/\/[^\s]+)/i;

export function isValidUrl(str?: string): boolean {
  return str ? !!URL_REGEX.test(str) : false;
}

const isFacebookGraphDependent = (url: string) => {
  const caseDesensitizedURL = url.toLowerCase();
  return (
    caseDesensitizedURL.includes('facebook.com') ||
    caseDesensitizedURL.includes('instagram.com')
  );
};

export const validOembedCheck = (embed: any, url: string) => {
  if (!isFacebookGraphDependent(url)) {
    if (embed?.html) {
      return true;
    }
  }
  return false;
};

export async function jsonFetch<T>(
  info: RequestInfo,
  init?: RequestInit
): Promise<T> {
  const res = await fetch(info, init);
  if (!res.ok) {
    throw new Error('Bad Fetch Response');
  }
  const data = await res.json();
  return data as T;
}

/**
 * Since there is no metadata persisted in a curio to determine what kind of
 * curio it is (Link or Text), this function determines by checking the
 * content's structure. Link curios are stored like so:
 * 
 * ['https://urbit.org']
 * 
 * Rich text is stored like so:
 * 
 * ['some text ', { bold: 'with style' }, ...]
 *
 * @param content CurioContent
 * @returns boolean
 */
export function isLinkCurio(content: CurioContent) {
  return content.length === 1 && typeof content[0] === 'string' && isValidUrl(content[0]);
}

export function linkFromCurioContent(content: CurioContent) {
  if(isLinkCurio(content)) {
    return content[0] as string;
  }

  return '';
}
