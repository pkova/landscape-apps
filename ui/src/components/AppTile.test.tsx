import React from 'react'
import { describe, expect, it } from 'vitest'
import renderer from 'react-test-renderer';
import { AppTile } from './AppTile'
import { render, screen } from '../../test/utils'

describe('AppTile', () => {
  describe('when image is available', () => {
    it('it renders an image', () => {
      render(<AppTile color='#fefefe' image="test.jpg" title="pals" />)
      expect(screen.getByAltText('app tile for pals')).toBeInTheDocument()
    })
  });

  describe('when image is not available', () => {
    it('it does not render an image', () => {
      render(<AppTile color='#fefefe' title="pals" />)
      expect(screen.queryByAltText('app tile for pals')).toBeNull()
    })
  });

  it('renders as expected', () => {
    const tree = renderer
      .create(<AppTile color='#fefefe' image="test.jpg" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})
