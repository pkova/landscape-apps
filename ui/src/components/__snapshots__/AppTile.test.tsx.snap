// Vitest Snapshot v1

exports[`AppTile > renders as expected 1`] = `
<div
  className="flex-none relative w-12 h-12 mr-3 rounded-lg bg-gray-200 overflow-hidden"
  style={
    {
      "backgroundColor": "#fefefe",
    }
  }
>
  <img
    alt="app tile for undefined"
    className="absolute top-0 left-0 h-full w-full object-cover"
    onError={[Function]}
    src="test.jpg"
  />
</div>
`;
